### DISCUSSION TITLE
Snippets (Blocks) from one file into a Layout File

### DISCUSSION THREAD

**Discussion by ValLynn** [11ty/eleventy#1881] [Q&A] [answered]
My objective is to separate the writing from the layout (to separate writing from HTML) but still be able to use Front Matter at the top of the document that holds the writing.  I'd like to be able to divide a document into "snippets" and insert those snippets into a layout template.  The layout template holds all of the html, the other file (the document), is written content (like a page in a journal).   I don't want to separate that document into multiple documents just to be able to insert those snippets into the HTML template (for example, Introduction.md, Body.md, Close.md) -  (see https://github.com/11ty/eleventy/issues/1467#issue-726693455).  

My earlier enthusiasm about extends and block was premature as there appears to be a limitation where we cannot use Front Matter with extends and blocks. I've made several goes at a Config file function workaround trying to follow (https://github.com/11ty/eleventy/issues/853#issuecomment-770751129) that ended up in nothing but undefined.  

Does anyone know of a workaround or something in Eleventy that I am missing that makes it possible to do what I want to do?

![1AA65A3B-7587-44E8-A5D7-C1B7645B22DA_1_201_a](https://user-images.githubusercontent.com/1017410/124632295-37377d00-de52-11eb-80ac-93bb9844f859.jpeg)

**Comment by pdehaan** [comment, marked as answer]
@ValLynn Do you have a sample repo with what you're trying to do?
I don't think I'm understanding how you're trying to separate front matter from layouts with blocks and snippets.

**Reply by ValLynn** [reply]
Thank you for responding.  My files are local only so I will need to come up with something to better communicate what I am trying to do.
.

**Reply by ValLynn** [reply]
Hopefully a picture is worth more than 1,000 words...

**Reply by pdehaan** [reply]
Yeah, thanks, that clears it up. I'll take a look and see if I have any Nunjucks block examples laying around. They can be a bit tricky sometimes as you have to use `{% extends "base.njk" %}` vs using `layout: base.njk` in front matter.

**Reply by ValLynn** [reply]
ummmm, you know I've read that a million times but it just now sunk in.  Now, I will delete the layout front matter in my file, add the extends, and see where that gets me.

**Reply by pdehaan** [reply]
Pretty old experiment, but might still be useful: https://github.com/pdehaan/11ty-blocks

**Reply by pdehaan** [reply]
New sample repo, but I'm having issues w/ having the .md template to render correctly.
https://github.com/pdehaan/11ty-blocks-extends

**Reply by ValLynn** [reply]
Ahhh  you sent me down the rabbit hole reviewing your **first** First Blood (hahaha) and in doing so, you've helped me find the workaround to my issue. I'm testing right now and will report in a few.  Also, your new repo - I'm using .njks, not .md. 

Next, based on this: 
> ...  use `{% extends "base.njk" %}` vs using `layout: base.njk` in front matter.

**I removed the front matter layout: declaration** (I hope that is the correct word) **from my .njk Content template file** (the one with the content (the left side of the image I uploaded)). 

This cleared up one problem and allowed the exposure of another problem (the front matter in the HTML Layout  file does not "feed" into the _includes that actually build the HTML Layout).  

BUT, after reviewing your repo, I am now putting that front matter into variables and I think I will get the functionality I am after. BRB

**Reply by ValLynn** [reply]
OMG - it is working.

**Reply by ValLynn** [reply]
> New sample repo, but I'm having issues w/ having the .md template to render correctly.
> https://github.com/pdehaan/11ty-blocks-extends

What is failing? (Also, I specified njk as my template enine in my .eleventy.js,  am using an .njk file as my content source and using {% extends %} just as you used it in your first repo (as a variable).

I had Front Matter in my HTML Layout file initially (this is not shown on the image I uploaded).  The HTML Layout file is just a bunch of partials I include. The partials are different parts of the structure (HTML) of the site such that once all of the partials are pulled in, the HTML is complete from `<!DOCTYPE ...>` to `</html>`. 

Before trying to use {% extends ...%} in the Content file (the left side of my image), the Front Matter in the HTML Layout file (an .njk) filled in the front matter placeholders in the partials (also .njk).  After using {% extends ... %}, filling in those placeholders with Front Matter from the HTML Layout file did not work ( @zachleat addressed this, I believe; I think this is the chaining he discusses here:  https://github.com/11ty/eleventy/issues/834#issuecomment-569474008 )

Note: prior to reading your post above, I had named the Layout file in the Front Matter of my Content template file AND I had an {% extends "layoutfilename.njk %} declaration in the content part too.

To get the Front Matter that used to be in the HTML Layout, I've just declared a couple of variables in a couple of files.  Now Snippeting works with my layout!  *THANK YOU!*

**Reply by pdehaan** [reply]
Your diagram said "Nunjucks or MD - I don't care which", so I tried with both (see src/about.md's generated [/www/about/index.html](https://github.com/pdehaan/11ty-blocks-extends/blob/main/www/about/index.html) file).

For some reason, when using a .md file with blocks, it's escaping some of the HTML elements (like doctype) and wrapping some other things in pre/code tags (which I can probably explain). But it just wasn't as easy as I was expecting, so I might be missing something somewhere.

Also, make sure to browse the docs for `{% extends %}` in the https://www.11ty.dev/docs/languages/nunjucks/ docs:

> `{% extends 'base.njk' %}` looks in `_includes/base.njk`. **Does not process front matter in the include file.**

**Reply by ValLynn** [reply]
> Also, make sure to browse the docs for `{% extends %}` in the https://www.11ty.dev/docs/languages/nunjucks/ docs:
> 
> > `{% extends 'base.njk' %}` looks in `_includes/base.njk`. **Does not process front matter in the include file.**

I did read that too (**after, of course**)

> For some reason, when using a .md file with blocks, it's escaping some of the HTML elements (like doctype)

Hmmm, I did not notice that.  Will test!

**Reply by pdehaan** [reply]
Yeah. Maybe I'm missing something with block layout w/ Markdown. It worked w/ njk, but I'm wondering if I need to use `|safe` with .md templates. I guess I just never tested that combination. Or else something about .md and markdown-it is escaping the doctype specifically and I need to fiddle some configs somewhere in the plumbing.

As for the rogue `<pre><code>` business, I assumed it was just https://www.11ty.dev/docs/languages/markdown/#there-are-extra-and-in-my-output and I need to disable the `code` rule.

I'd say stick to .njk templates, and maybe use some other method of parsing markdown (custom filters or paired shortcodes) instead.

**Reply by ValLynn** [reply]
> wrapping some other things in pre/code tags

I just took my .njk content file, duplicated it, stripped it to markdown and extend plus block and yes, I', getting the weird pre/code tags and escape stuff also! If that had happened before, I had not noticed it.

*BUT* in my case, this is not coming from the markdown file itself but appears to be a result of using the markdown file in the entire "cascade." All of the other files (the Layout file and the _includes are .njk). The presence of _that one markdown file_, even though **it** does not have any indents in it, causes indents in any of the other files to render with the pre and code tags. Huh... whaddyaknow...

When I limit my indents to no more than 4 spaces, the pre and code gobbledeegoop goes away.

**Reply by ValLynn** [reply]
> I'd say stick to .njk templates, and maybe use some other method of parsing markdown (custom filters or paired shortcodes) instead.

This morning I would have agreed with you but by golly, I've got a Markdown file with an {% extends %} and  2 {% blocks %}  that are being pulled correctly into my HTML .njk layout file. EU-effing-REKA! AND Front Matter!  Working with Front Matter!

**Reply by ValLynn** [reply]
Thank you @pdehaan. I've learned quite a bit today because of you and this discussion.

**Reply by pdehaan** [reply]
Yeah, I fixed `<pre><code>` (due to indenting 4+ spaces), with the following:

```diff
  const markdownLib = markdownIt({
    html: true,
    break: false,
    linkify: true,
  })
+ .disable("code");
  eleventyConfig.setLibrary("md", markdownLib);
```

I pushed a newer version to https://github.com/pdehaan/11ty-blocks-extends but still no luck on the nesting of `<p>` tags in the Markdown version. I found a few related issues in markdown-it repo, but they were closed as "not a bug". So I think I'm giving up on debugging `{%block%}` tags in .md files directly and will just stick w/ .njk or .html templates for blocks.

I also added examples of a "md" filter and paired shortcode if you wanted to still use markdown in your blocks.

**Reply by pdehaan** [reply]
> This morning I would have agreed with you but by golly, I've got a Markdown file with an {% extends %} and 2 {% blocks %} that are being pulled correctly into my HTML .njk layout file. EU-effing-REKA! AND Front Matter! Working with Front Matter!

Let me know if you ever post this to a public repo. I'd love to check it out.

And to think, it all started with a paper repo(tm).

**Reply by ValLynn** [reply]
Ha Ha ha my paper repo worked!  

I will take a look at the changes to your repo.  I am very new to Eleventy and what ad hoc Javascript skill I had went by the wayside in 2013 - so I'm learning Javascript as I go.  

I'll reach out when I post the files. That will be yet another learning opportunity!

**Reply by ValLynn** [reply]
> Let me know if you ever post this to a public repo.

New repo  --> https://github.com/ValLynn/11ty-blocks-extends

**Reply by pdehaan** [reply]
That's great!
Now I need to figure out why your Markdown template works so well, but mine was injecting some extra `<p>` tags.
